opencti-platform/opencti-front/src/private/components/workspaces/investigations/InvestigationExpandForm.tsx

```diff
@@ -0,0 +1,403 @@
+  const [existingRels, setExistingRels] = useState<Map<string, Map<string, string[]>>>(new Map());
+
+  // How many entity types and rel types we already have in the graph.
+  useEffect(() => {
+    const countTargets = new Map<string, Map<string, number>>();
+    const countRels = new Map<string, Map<string, string[]>>();
+
+    links.forEach((link) => {
+      // toLowerCase() because relationship names are pascalized
+      const relType = link.entity_type.toLowerCase();
+
+      const from = link.source.id;
+      const to = link.target.id;
+      // Init Maps if required
+      if (from && !countTargets.has(from)) {
+        countTargets.set(from, new Map());
+      }
+      if (to && !countTargets.has(to)) {
+        countTargets.set(to, new Map());
+      }
+      if (from && !countRels.has(from)) {
+        countRels.set(from, new Map());
+      }
+      if (to && !countRels.has(to)) {
+        countRels.set(to, new Map());
+      }
+
+      // toLowerCase() because relationship names are pascalized
+      // in elAggregationRelationsCount().
+      const typeFrom = (link.target.entity_type ?? '').toLowerCase();
+      const targetFrom = countTargets.get(from);
+      if (targetFrom) {
+        const currentTarget = targetFrom?.get(typeFrom);
+        if (currentTarget !== undefined) {
+          targetFrom.set(typeFrom, currentTarget + 1);
+        } else {
+          targetFrom.set(typeFrom, 1);
+        }
+      }
+      const relFrom = countRels.get(from);
+      if (relFrom) {
+        const currentRel = relFrom?.get(relType);
+        if (currentRel !== undefined) {
+          relFrom.set(relType, [...currentRel, link.id]);
+        } else {
+          relFrom.set(relType, [link.id]);
+        }
+      }
+
+      // toLowerCase() because relationship names are pascalized
+      // in elAggregationRelationsCount().
+      const typeTo = (link.source.entity_type ?? '').toLowerCase();
+      const targetTo = countTargets.get(to);
+      if (targetTo) {
+        const current = targetTo?.get(typeTo);
+        if (current !== undefined) {
+          targetTo.set(typeTo, current + 1);
+        } else {
+          targetTo.set(typeTo, 1);
+        }
+      }
+      const relTo = countRels.get(to);
+      if (relTo) {
+        const currentRel = relTo?.get(relType);
+        if (currentRel !== undefined) {
+          relTo.set(relType, [...currentRel, link.id]);
+        } else {
+          relTo.set(relType, [link.id]);
+        }
+      }
+    });
+
+    setExistingTargets(countTargets);
+    setExistingRels(countRels);
+  }, [links, setExistingRels, setExistingTargets]);
+
+  // Fetch all possible relationships linked to the nodes to expand.
+  useEffect(() => {
+    async function fetchRelationships() {
+      if (selectedNodes.length === 0) return;
+      const nodeIds = Array.from(selectedNodes).map((node) => node.id);
+
+      const { stixRelationshipsDistribution } = await fetchQuery(
```

Comment:
I don't think this is the best way of doing it. We can discuss it if you want but yo should use a relay hook to perform a query from a ref already declared